api/ngrok/v1alpha1/agentendpoint_types.go

```diff
@@ -0,0 +1,169 @@
+	// Scheme (shorthand) - https:// or tcp:// or tls:// or http://
+	//     When using scheme you are defining the protocol and will receive back a randomly assigned ngrok address.
+	// Empty - ``
+	//     When empty your endpoint will default to be https and receive back a randomly assigned ngrok address.
+	// Internal - some.domain.internal
+	//     When ending your url with .internal, an internal endpoint will be created. nternal Endpoints cannot be accessed directly, but rather
+	//     can only be accessed using the forward-internal traffic policy action.
+	//
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+
+	// Allows configuring a TrafficPolicy to be used with this AgentEndpoint
+	//
+	// +kubebuilder:validation:Optional
+	TrafficPolicy *TrafficPolicyCfg `json:"trafficPolicy,omitempty"`
+
+	// Human-readable description of this agent endpoint
+	//
+	// +kubebuilder:default:=`Created by the ngrok-operator`
+	Description string `json:"description,omitempty"`
+
+	// String of arbitrary data associated with the object in the ngrok API/Dashboard
+	//
+	// +kubebuilder:default:=`{"owned-by":"ngrok-operator"}`
+	Metadata string `json:"metadata,omitempty"`
+
+	// List of Binding IDs to associate with the endpoint
+	// Accepted values are "public", "internal", or strings matching the pattern "k8s/*"
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Items=pattern=`^(public|internal|k8s/.*)$`
+	Bindings []string `json:"bindings,omitempty"`
+}
+
+type TrafficPolicyCfgType string
+
+const (
+	TrafficPolicyCfgType_K8sRef TrafficPolicyCfgType = "targetRef"
+	TrafficPolicyCfgType_Inline TrafficPolicyCfgType = "inline"
+)
+
+func (t TrafficPolicyCfgType) IsKnown() bool {
+	switch t {
+	case TrafficPolicyCfgType_K8sRef, TrafficPolicyCfgType_Inline:
+		return true
+	default:
+		return false
+	}
+}
+
+// +kubebuilder:validation:XValidation:rule="self.type == 'inline' ? has(self.inline) && !has(self.targetRef) : true",message="When type is inline, inline must be set, and targetRef must not be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'targetRef' ? has(self.targetRef) && !has(self.inline) : true",message="When type is targetRef, targetRef must be set, and inline must not be set."
+type TrafficPolicyCfg struct {
+	// Controls the way that the TrafficPolicy configuration will be provided to the Agent Endpoint
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=targetRef;inline
+	Type TrafficPolicyCfgType `json:"type"`
+
+	// Inline definition of a TrafficPolicy to attach to the agent Endpoint
+	// The raw json encoded policy that was applied to the ngrok API
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:Type=object
+	Inline json.RawMessage `json:"inline,omitempty"`
```

Comment:
IMO Switching from raw json message to structural types would be a good idea, but for now I want to keep consistency with the other resources. We can discuss whether we want to change this prior to GA in a follow-up.